signature/signature_test.go

```diff
@@ -1,253 +1,184 @@
+			receivedAt: "2021-07-05T12:00:00+02:00",
+			signature:  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJNZXNzYWdlQmlyZCIsImlhdCI6MTYyNTQ3OTIwMCwiZXhwIjoxNjI1NDc5MjYwLCJqdGkiOiIyNDNjMjdhZS0yZjAyLTQ2YTAtODg1Mi1jNjZmMzdlYTlmNDYiLCJ1cmxfaGFzaCI6IjBmMTE1ZGIwNjJiN2MwZGQwMzBiMTY4NzhjOTlkZWE1YzM1NGI0OWRjMzdiMzhlYjg4NDYxNzljNzc4M2U5ZDcifQ._Uwf4HMtfAT6jvbBbh85Q9TunX0QlsXoaLGKX0I4VDg",
+			wantErr:    "invalid jwt: signature is invalid",
 		},
+
 		{
-			name: "Request with wrong signature header",
-			k:    testKey,
-			ts:   testTs,
-			s:    testSignature,
-			sh:   "wrong-header",
-			tsh:  tsHeader,
-			e:    http.StatusUnauthorized,
+			name:       "invalid missing payload",
+			receivedAt: "2021-07-05T12:00:00+02:00",
+			signature:  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJNZXNzYWdlQmlyZCIsImlhdCI6MTYyNTQ3OTIwMCwiZXhwIjoxNjI1NDc5MjYwLCJqdGkiOiIxNDUwMTUzMi05NmYyLTQ2ODQtOTgzMi02OGYwOTUxYWUzNDIiLCJ1cmxfaGFzaCI6IjBmMTE1ZGIwNjJiN2MwZGQwMzBiMTY4NzhjOTlkZWE1YzM1NGI0OWRjMzdiMzhlYjg4NDYxNzljNzc4M2U5ZDciLCJwYXlsb2FkX2hhc2giOiIzMjRjYzA2N2IyNTdlZGEwYmNiZDljOGQ4MTgwNzdhMDlhOTU2OGMwZDRjYTA2MDM4ZGVkOGZhZGRmODEzZmQ2In0.rQqiANogDOMafgg_B6p362PuhInAro9lMm2j_vruBA0",
+			wantErr:    "invalid jwt: payload_hash was set; expected no payload value",
 		},
+
 		{
-			name: "Request with wrong timestamp header",
-			k:    testKey,
-			ts:   testTs,
-			s:    testSignature,
-			sh:   sHeader,
-			tsh:  "wrong-header",
-			e:    http.StatusUnauthorized,
+			name:           "invalid unexpected payload",
+			requestPayload: "Hello, World!",
+			receivedAt:     "2021-07-05T12:00:00+02:00",
+			signature:      "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJNZXNzYWdlQmlyZCIsImlhdCI6MTYyNTQ3OTIwMCwiZXhwIjoxNjI1NDc5MjYwLCJqdGkiOiI1OWEyNDRkYy1lOWFkLTRlMjMtOTc3OC0zNzFmYWEyMzhmNzIiLCJ1cmxfaGFzaCI6IjBmMTE1ZGIwNjJiN2MwZGQwMzBiMTY4NzhjOTlkZWE1YzM1NGI0OWRjMzdiMzhlYjg4NDYxNzljNzc4M2U5ZDcifQ.SrhlKJ-ES4Dg8BBXKtop3u92Z_k4L4VjHKsyHWpweGE",
+			wantErr:        "invalid jwt: payload_hash is invalid",
 		},
 	}
 
-	for _, tt := range cases {
-		v := NewValidator(tt.k)
-		testTime, _ := stringToTime(testTs)
-		ValidityWindow = time.Now().Add(time.Second*1).Sub(testTime) * 2
-		ts := httptest.NewServer(v.Validate(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-			w.WriteHeader(http.StatusOK)
-		})))
-		defer ts.Close()
-
-		client := &http.Client{}
-		req, _ := http.NewRequest("GET", ts.URL+"?"+testQp, strings.NewReader(testBody))
-		req.Header.Set(tt.sh, tt.s)
-		req.Header.Set(tt.tsh, tt.ts)
-		res, _ := client.Do(req)
-		assert.Equal(t, tt.e, res.StatusCode)
-	}
+	for _, test := range cases {
+		t.Run(test.name, func(t *testing.T) {
+			TimeFunc = func() time.Time {
+				r, _ := time.Parse(time.RFC3339, test.receivedAt)
+				return r
+			}
 
+			v := NewValidator(testSecret)
+			reqUrl := testBaseUrl + test.requestParams
+			if test.requestParams == "" {
+				reqUrl += "/"
+			}
+			err := v.ValidSignature(test.signature, reqUrl, []byte(test.requestPayload))
```

Comment:
perhaps `ValidSignature` should take a timestamp and an `http.Request` instead of these parameters? That would make it easier for the user since when they're calling, they have both.

Reply:
This is the unit test for `ValidSignature`, there's also `ValidRequest` method that takes `http.Request` instead of these params. 
For the timestamp, I believe it should only be the current timestamp because the only use-case for a user to have a different timestamp is when they do testing. In that case, it can be changed by overriding the TimeFunc method.